platform/fabric/core/generic/committer/em.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package committer
+
+import (
+	"context"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/fabric/driver"
+	pb "github.com/hyperledger/fabric-protos-go/peer"
+)
+
+// EventManager manages events for the commit pipeline.
+// It consists of a central queue of events.
+// The queue is fed by multiple sources.
+// A single thread reads from this queue and invokes the listeners in a blocking way
+type EventManager struct {
+	EventQueue chan TxEvent
+	Vault      driver.Vault
+
+	allListeners  []driver.TxStatusListener
+	txIDListeners map[string][]driver.TxStatusListener
+	mutex         sync.RWMutex
+}
+
+func NewEventManager(vault driver.Vault, size int) *EventManager {
+	return &EventManager{
+		EventQueue:    make(chan TxEvent, size),
+		Vault:         vault,
+		allListeners:  nil,
+		txIDListeners: map[string][]driver.TxStatusListener{},
+	}
+}
+
+func (c *EventManager) AddListener(txID string, l driver.TxStatusListener) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if len(txID) == 0 {
+		c.allListeners = append(c.allListeners, l)
+	}
+
+	ls, ok := c.txIDListeners[txID]
+	if !ok {
+		ls = []driver.TxStatusListener{}
+		c.txIDListeners[txID] = ls
+	}
+	ls = append(ls, l)
+	c.txIDListeners[txID] = ls
+}
+
+func (c *EventManager) DeleteListener(txID string, ch driver.TxStatusListener) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	var ls []driver.TxStatusListener
+	if len(txID) == 0 {
+		ls = c.allListeners
+	} else {
+		var ok bool
+		ls, ok = c.txIDListeners[txID]
+		if !ok {
+			return
+		}
+	}
+	for i, l := range ls {
+		if l == ch {
+			ls = append(ls[:i], ls[i+1:]...)
+			c.txIDListeners[txID] = ls
+			return
+		}
+	}
+}
+
+func (c *EventManager) Post(event TxEvent) {
+	c.EventQueue <- event
+}
+
+func (c *EventManager) Dispatch(event TxEvent) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("caught panic while running dispatching event [%v]: [%s][%s]", event, r, debug.Stack())
+		}
+	}()
+	l := c.cloneListeners(event.TxID)
+	for _, listener := range l {
+		if err := listener.OnStatus(event.TxID, int(event.ValidationCode), event.ValidationMessage); err != nil {
+			logger.Errorf("failed on status call [%v]: [%s][%s]", event, err, debug.Stack())
+		}
+		c.DeleteListener(event.TxID, listener)
```

Comment:
You could delete all listeners the moment you clone them, right? We don't have any revert mechanism and we keep the write lock for O(n^2) time like this while we can do it in O(1).